client/client.go

```diff
@@ -373,28 +402,16 @@ func (r *NotaryRepository) Publish() error {
 		return err
 	}
 
-	// ensure we can marshal all the json before sending anything to remote
-	targetsJSON, err := json.Marshal(targets)
-	if err != nil {
-		return err
-	}
-	snapshotJSON, err := json.Marshal(snapshot)
-	if err != nil {
-		return err
-	}
-	update := make(map[string][]byte)
-	// if we need to update the root, marshal it and push the update to remote
-	if updateRoot {
-		rootJSON, err := json.Marshal(root)
-		if err != nil {
-			return err
-		}
-		update["root"] = rootJSON
-	}
-	update["targets"] = targetsJSON
-	update["snapshot"] = snapshotJSON
 	err = remote.SetMultiMeta(update)
 	if err != nil {
+		// TODO: this isn't exactly right, since there could be lots of
+		// reasons a request 400'ed.  Need better error translation from HTTP
+		// status codes maybe back to the server errors?
+		if _, ok := err.(store.ErrInvalidOperation); ok && clientCantSignSnapshot {
```

Comment:
Hmm... the server would only be able to send something like BadHierarchy, because it doesn't know if it's supposed to have the key and just doesn't.  And Publish doesn't know if it's supposed to be the signer either, because that is missing information that is inferred - shouldn't it really be the same error as the user just missing the key (e.g. `ErrNoKeys`)?